File: 111-minimum-depth-of-binary-tree.py
# ❓ DESCRIPTION:
# Given a binary tree, find its minimum depth.
# The minimum depth is the number of nodes along the shortest path from the root node down to the nearest leaf node.
# Note: A leaf is a node with no children.

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

# ✅ SOLUTION:
class Solution(object):
    def minDepth(self, root):
        if root is None:  return 0

        leftDepth = self.minDepth(root.left)
        rightDepth = self.minDepth(root.right)

        if root.left is None and root.right is None:
            return 1
       
        if root.left is None:
            return 1 + rightDepth
       
        if root.right is None:
            return 1 + leftDepth
       
        return min(leftDepth, rightDepth) + 1;  

File: 111-minimum-depth-of-binary-tree.js
// Definition for a binary tree node.
function TreeNode(val , left , right) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
}

/**
 * @param {TreeNode} root
 * @return {number}
 */
const minDepth = function (root) {
    if (!root) return 0;
    let depth = 1;

    const queue = [root];

    while (queue.length) {

        const levelSize = queue.length;

        for (let i = 0; i < levelSize; i++) {
            const node = queue.shift();

            if (!node.left && !node.right) return depth;

            if (node.left) queue.push(node.left);
            if (node.right) queue.push(node.right);
        }
        depth++;
    }
};

const minDepth = function (root) {
    if (root == null) return 0;
    if (root.left == null && root.right == null)
        return 1;
    if (root.left == null)
        return 1 + minDepth(root.right);
    if (root.right == null)
        return 1 + minDepth(root.left);
    return Math.min(minDepth(root.left) , minDepth(root.right)) + 1;
};

console.log(minDepth([]));
console.log(
    minDepth(
        new TreeNode(
            3 ,
            new TreeNode(9) ,
            new TreeNode(20 , new TreeNode(15) , new TreeNode(7)) ,
        ) ,
    ) ,
);
console.log(
    minDepth(
        new TreeNode(
            2 ,
            null ,
            new TreeNode(
                3 ,
                null ,
                new TreeNode(4 , null , new TreeNode(5 , null , new TreeNode(6))) ,
            ) ,
        ) ,
    ) ,
);
